x-pack/plugins/apm/server/routes/storage_explorer/get_storage_details_per_service.ts

```diff
@@ -153,3 +155,122 @@ export async function getStorageDetailsPerProcessorEvent({
     };
   });
 }
+
+export async function getStorageDetailsPerIndex({
+  setup,
+  context,
+  indexLifecyclePhase,
+  randomSampler,
+  start,
+  end,
+  environment,
+  kuery,
+  serviceName,
+}: {
+  setup: Setup;
+  context: ApmPluginRequestHandlerContext;
+  indexLifecyclePhase: IndexLifecyclePhaseSelectOption;
+  randomSampler: RandomSampler;
+  start: number;
+  end: number;
+  environment: string;
+  kuery: string;
+  serviceName: string;
+}) {
+  const { apmEventClient } = setup;
+
+  const [
+    { indices: allIndicesStats },
+    indicesLifecycleStatus,
+    indicesInfo,
+    response,
+  ] = await Promise.all([
+    getTotalIndicesStats({ setup, context }),
+    getIndicesLifecycleStatus({ setup, context }),
+    getIndicesInfo({ setup, context }),
+    apmEventClient.search('get_storage_details_per_index', {
+      apm: {
+        events: [
+          ProcessorEvent.span,
+          ProcessorEvent.transaction,
+          ProcessorEvent.error,
+          ProcessorEvent.metric,
+        ],
+      },
+      body: {
+        size: 0,
+        track_total_hits: false,
+        query: {
+          bool: {
+            filter: [
+              ...environmentQuery(environment),
+              ...kqlQuery(kuery),
+              ...rangeQuery(start, end),
+              ...termQuery(SERVICE_NAME, serviceName),
+              ...(indexLifecyclePhase !== IndexLifecyclePhaseSelectOption.All
+                ? termQuery(
+                    TIER,
+                    indexLifeCyclePhaseToDataTier[indexLifecyclePhase]
+                  )
+                : []),
+            ] as QueryDslQueryContainer[],
+          },
+        },
+        aggs: {
+          sample: {
+            random_sampler: randomSampler,
+            aggs: {
+              indices: {
+                terms: {
+                  field: INDEX,
+                  size: 500,
+                },
+                aggs: {
+                  number_of_metric_docs_for_index: {
+                    value_count: {
+                      field: INDEX,
+                    },
+                  },
+                },
+              },
+            },
+          },
+        },
+      },
+    }),
+  ]);
+
+  return (
+    response.aggregations?.sample.indices.buckets.map((bucket) => {
+      const indexName = bucket.key as string;
+      const numberOfDocs = bucket.number_of_metric_docs_for_index.value;
+      const indexInfo = indicesInfo[indexName];
+      const indexLifecycle = indicesLifecycleStatus[indexName];
+
+      const size =
+        allIndicesStats &&
+        getEstimatedSizeForDocumentsInIndex({
+          allIndicesStats,
+          indexName,
+          numberOfDocs,
+        });
+
+      return {
+        indexName,
+        numberOfDocs,
+        primary: indexInfo
+          ? indexInfo?.settings?.index?.number_of_shards ?? 0
```

Comment:
You ca remove the first `?` since you're already checking it. The same thing for `replica`
```suggestion
        primary: indexInfo
          ? indexInfo.settings?.index?.number_of_shards ?? 0
```

Reply:
Good spot, removed it in https://github.com/elastic/kibana/pull/143179/commits/3eb643d07c1780558b5b78e39bd8776e57540e09